### ISSUE TITLE
Type casting on Include.on option for Model.findAll()

### ISSUE THREAD

**Issue by bosunolanrewaju** [sequelize/sequelize#8574]
## What are you doing?

Specify a custom join with `include.on` to eager records from an associated model. Mainmodel column `id` is INTEGER type while ModelB column `key` is TEXT type on the db.
```js
// Query on Mainmodel
{ 
    where: { . . . },
    include: [ModelA, {
        model: ModelB,
        on: {
            "$modelb.key$": { $col: "Mainmodel.id" },
        },
    }],
}
```

## What do you expect to happen?
_I expect to have result which includes ModelB data where key is the same as id on Mainmodel_

## What is actually happening?
_I got an error_

_SequelizeDatabaseError: operator does not exist: text = integer_

How do I do casting is this scenario?

__Dialect:__ postgres
__Dialect version:__ 9.6.3
__Database version:__ 9.6.3
__Sequelize version:__ 4.4.2
__Tested with master branch:__ Yes

**Note :** _Your issue may be ignored by maintainers if it's not tested against `master` branch OR does not follow issue template._

**Comment by iamjochem** [comment]
hi,

there is `sequelize.cast()` which *should* be able to help you, does the following get you anywhere? (you may have to play around with the exact "type" argument): 

```js
{ 
    where: { . . . },
    include: [ModelA, {
        model: ModelB,
        on: {
            "$modelb.key$": sequelize.cast({ $col: "Mainmodel.id" }, 'text'),
        },
    }],
}
```

**Comment by bosunolanrewaju** [comment]
@iamjochem Thank you for your response. So, I tried your suggestion and it came back with this error:

```
TypeError: s.replace is not a function
   at Object.removeTicks (/server/node_modules/sequelize/lib/utils.js:415:12)
   at Object.quoteIdentifier (/server/node_modules/sequelize/lib/dialects/postgres/query-generator.js:854:35)
   at Object._getSafeKey (/server/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2285:33)
   at Object._joinKeyValue (/server/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2261:16)
   at Object.whereItemQuery (/server/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2091:17)
   at Utils.getComplexKeys.forEach.prop (/server/node_modules/sequelize/lib/dialects/abstract/query-generator.js:1944:25)
   at Array.forEach (<anonymous>)
```

Apparently, this `sequelize.cast({ $col: "Mainmodel.id" }, 'text')` returns a sequelize symbol and not a String. So `.replace()` is not available.

Any help to figure this out would be appreciated. At the moment, I'm using raw sql to run the query on the db, but I will like to have sequelize handle it if possible.

**Comment by iamjochem** [comment]
:-(

the Symbol is come from relatively new functionality, it replaces old '$bla' style query building logic with something more structured/explicit (instances of `SequelizeMethod`) but there is also logic to automatically map the "old" strings to the new underlying Symbols ... I'd hazard a guess that maybe the implementation is still a little rought around the edges (the reason this was all changed is related to security)

you could try this (pure guesswork on my part, you might get lucky but chances are you'll run into the same error):

```js
{ 
    where: { . . . },
    include: [ModelA, {
        model: ModelB,
        on: {
            "$modelb.key$": sequelize.cast(sequelize.col("Mainmodel.id"), 'text'),
        },
    }],
}
```

if you luck out then it's seems you've run into a bug,  it _looks_ like [_getSafeKey()](https://github.com/sequelize/sequelize/blob/e40f6f45981db87bdf4c62a93987d220cab2cfd5/lib/dialects/abstract/query-generator.js#L2265) is not taking into account the siutation where the `key` is a symbol (which is the half-way variant between "normal attribute name" and the new SequelizeMethod functionality) ... looking at the code I find it difficult to determine whether this situation should be accounted for in this particular place in the code

**Comment by iamjochem** [comment]
looks like the problem has been fixed by the changes in #8068
